src/cfront.c

```diff
@@ -678,7 +722,7 @@ void lex_expect(token_t token)
     next_token = get_next_token();
 }
 
-void read_expr(int param_no, block_t *parent);
+void read_expr(int param_no, block_t *parent, macro_t *macro);
```

Comment:
Don't do that. Avoid changing the function prototypes. The macros should be regarded as a special case of expressions.

Reply:
We need to check the variables are live in that block or not. The macro is in the same way, but it doesn't need to know how the variable is exactly. The only thing it should know is where the argument from (the position in the source code).

In practice, the arguments of the macro have the higher priority to search that cause the behavior different from the normal variables. We search in the scope of the macro first, then the block. Or, we can say the scope of a macro is the **superset** of the block it belonging.

Reply:
> Or, we can say the scope of a macro is the **superset** of the block it belonging.

I agree with this concept. Let's unify within the pull request.